Fix for #1916

jaloren opened the pull request:
Fix for #1916

robotci commented:
Admin, can CI be enabled?

pekkaklarck commented on atest/robot/test_libraries/logging_api.robot:
```diff
@@ -15,9 +15,19 @@ Log levels
+
+Log Error Using Print
+    ${tc}=    Check test case    ${TEST NAME}
+    Check log message  ${tc.kws[0].msgs[0]}   error msg with print   ERROR
```
This test suite is for testing `robot.api.logger`. Tests for using print/stdout belong to `print_logging.robot` suite instead. Also, just one empty row between tests.

pekkaklarck commented on atest/robot/test_libraries/logging_with_logging.robot:
```diff
@@ -16,18 +16,18 @@ Log with default levels
     Check log message  ${tc.kws[0].msgs[2]}  between debug and info    DEBUG
     Check log message  ${tc.kws[0].msgs[3]}  between info and warning  INFO
-    Check log message  ${tc.kws[0].msgs[4]}  above warning             WARN
+    Check log message  ${tc.kws[0].msgs[4]}  above warning             ERROR
```
It would be good to test also a custom level between warning and error. That should be mapped to WARN because levels between info and warning are mapped to INFO. Anything above error should be ERROR.

pekkaklarck commented on atest/testdata/test_libraries/LibUsingLoggingApi.py:
```diff
@@ -24,9 +24,17 @@ def write_messages_to_console():
     logger.console('in two parts')
     logger.info('To log and console', also_console=True)
 
+
```
Adding an empty line here is OK because PEP-8 recommends two empty lines between top-level functions. They should be added consistently in the whole module, though. Using just one empty line here is fine too.

pekkaklarck commented on atest/testdata/test_libraries/LibUsingLoggingApi.py:
```diff
@@ -24,9 +24,17 @@ def write_messages_to_console():
+
+
+def log_error_with_print():
+    print "*ERROR* error msg with print"
```
As already commented, this suite/library is about `robot.api.logger`, not about logging using `print`.

pekkaklarck commented on doc/userguide/src/ExtendingRobotFramework/CreatingTestLibraries.rst:
```diff
@@ -1189,7 +1189,19 @@ messages, specify the log level explicitly by embedding the level into
-`INFO`, `WARN`, `FAIL` and `HTML`.
+`INFO`, `WARN`, `FAIL`, `ERROR` and `HTML`.
+
+
```
One empty line between sections is enough.

pekkaklarck commented on doc/userguide/src/ExtendingRobotFramework/CreatingTestLibraries.rst:
```diff
@@ -1189,7 +1189,19 @@ messages, specify the log level explicitly by embedding the level into
+using them for reporting important but non-critical problems to users.
+
+__ `Errors and warnings during execution`_
+
```
This is identical documentation as with `Warnings` below, right? Why not just have one section `Warnings and errors` instead? Should also add a note that with errors this functionality is new in 2.9.

pekkaklarck commented on doc/userguide/src/ExtendingRobotFramework/CreatingTestLibraries.rst:
```diff
@@ -1424,9 +1445,9 @@ Framework.
+Robot Framework. Its levels `DEBUG`, `ERROR`, and `INFO` are mapped
 directly to the matching Robot Framework log levels and `WARNING`
-and everything above is mapped to `WARN`. Custom levels below
+and `CRITICAL` are mapped to `WARN`. Custom levels below
```
Is `CRITICAL` really mapped to `WARN`? Isn't it above `ERROR` and thus mapped to `ERROR`?

pekkaklarck commented on src/robot/libraries/BuiltIn.py:
```diff
@@ -2134,12 +2134,12 @@ def catenate(self, *items):
         u"""Logs the given message with the given level.
 
-        Valid levels are TRACE, DEBUG, INFO (default), HTML, and WARN.
+        Valid levels are TRACE, DEBUG, INFO (default), HTML, ERROR, and WARN.
```
I'd prefer ERROR after WARN because it is higher.

pekkaklarck commented on src/robot/output/logger.py:
```diff
@@ -116,7 +116,7 @@ def _log_message(self, msg):
         for logger in self._loggers.all_loggers():
             logger.log_message(msg)
-        if msg.level == 'WARN':
+        if msg.level == 'WARN' or msg.level == 'ERROR':
```
`msg.level in ('WARN', 'ERROR')` reads a bit better than the above. According to `timeit` it is slightly faster too.

pekkaklarck replied:
This I'd still like to be changed.

pekkaklarck commented on src/robot/output/pyloggingconf.py:
```diff
@@ -22,6 +22,7 @@
           'DEBUG': logging.DEBUG,
+          'ERROR': logging.ERROR,
           'INFO': logging.INFO,
           'WARN': logging.WARNING}
```
Would be better to have these in priority order. In other words, ERROR should be last.

pekkaklarck commented on src/robot/reporting/jsmodelbuilders.py:
```diff
@@ -146,6 +146,8 @@ class MessageBuilder(_Builder):
         if msg.level == 'WARN':
             self._context.create_link_target(msg)
+        if msg.level == 'ERROR':
+             self._context.create_link_target(msg)
```
Just one `if` with `msg.level in ('WARN', 'ERROR')` is enough here.

pekkaklarck commented:
This PR looks very good in general. I added quite many line notes, but they should all be pretty easy to fix. Let me know if you have any questions about them.

pekkaklarck commented:
@robotci enable

jaloren commented:
I believe I incorporated all the feedback that you gave. Thanks!!

In regards to the test_jsmodelbuilders.py unit test, you were correct. It is passing before any of my commits. However, after my commits, three unit tests are failing. The failure information is a bit opaque so it may take me a little while to figure  why the test cases are failing, how to fix them, and then of course add a unit test for the functionality I implemented.

pekkaklarck commented:
Let me know what tests are failing and what error you get if you cannot easily figure out why they fail.

pekkaklarck commented on atest/robot/test_libraries/logging_api.robot:
```diff
@@ -15,9 +15,13 @@ Log levels
-
+    Check log message  ${ERRORS.msgs[2]}     error msg 1   ERROR
+    Check log message  ${ERRORS.msgs[3]}     error msg 2   ERROR 
+       
```
There are apparently trailing spaces here. They should be avoided. I generally recommend setting IDE to remove them automatically. (This reminds me that I should be writing contribution guidelines where this stuff is mentioned.)

jaloren replied:
Noted. I actually am a big fan of the linter Brian Oakely wrote https://github.com/boakley/robotframework-lint. It would be pretty nifty if there was a simple script i could run that would flag anything that doesn't conform to whatever guidelines your project defines.

jaloren replied:
What IDE do you use? I have a mac and have had a hellacious time getting ride to work because wxpython packaging is a nightmare.

guykisel replied:
This might sound silly, but when I develop on a mac I use a linux VM to run RIDE.

pekkaklarck commented on atest/testdata/test_libraries/LibUsingLoggingApi.py:
```diff
@@ -24,9 +24,13 @@ def write_messages_to_console():
 def log_callable():
     logger.info(log_callable)
+    
+
```
Empty rows at the end should be avoided. They add noise to diffs. The last line should end with a newline, though. (Again something that IDEs can do automatically and something that should be mentioned in contribution guidelines.)

pekkaklarck commented on atest/testdata/test_libraries/logging_api.robot:
```diff
@@ -25,3 +25,7 @@ Log Non-Strings
+
+
+Log Error Using Print
+    Log Error With Print
```
Is this a leftover from earlier test? Doesn't belong to this suite anyway.

pekkaklarck commented on atest/testdata/test_libraries/print_logging.robot:
```diff
@@ -29,3 +29,8 @@ Logging HTML
+Log Error With Print
+    Print Error   This is an error msg logged with print.
+
+    
```
Extra empty lines at the end.

pekkaklarck commented on atest/testresources/testlibs/ExampleLibrary.py:
```diff
@@ -167,6 +167,10 @@ def __unicode__(self):
+        print '*ERROR* {0}'.format(msg)
+        
+
 
```
Two empty lines is enough here.

pekkaklarck commented on doc/userguide/src/ExtendingRobotFramework/CreatingTestLibraries.rst:
```diff
@@ -1189,16 +1189,19 @@ messages, specify the log level explicitly by embedding the level into
+files. This makes errors more visible than other messages and allows
 using them for reporting important but non-critical problems to users.
 
+.. note:: Use of the `ERROR` log level is new in Robot 2.9.
```
ERROR level itself isn't new in 2.9. Should probably be something like "Automatically adding ERRORs logged by keywords to the Test Execution Errors section is new in Robot Framework 2.9.".

pekkaklarck commented on doc/userguide/src/ExtendingRobotFramework/CreatingTestLibraries.rst:
```diff
@@ -1424,11 +1433,10 @@ Framework.
+Robot Framework. Its levels `DEBUG`, `ERROR`, and `INFO` are mapped
+directly to the matching Robot Framework log levels. The `WARNING`
+log level is mapped to `WARN` and `CRITICAL` is mapped to `ERROR`. Custom levels
+above `ERROR` are mapped to `ERROR` Custom levels below `DEBUG` are mapped to `DEBUG` and everything between `DEBUG` and `WARNING` is mapped to `INFO`.
```
The original documentation was already a bit complicated and adding ERROR into the mix makes it worse. Perhaps it would be better to say something like this:

> Its levels `DEBUG` , `INFO`, `WARNING` and `ERROR` are mapped directly to the matching Robot Framework log levels and `CRITICAL` is mapped to `ERROR. Possible custom levels are mapped to the closest standard level. For example, a custom level between `INFO` and `WARNING` is consider `INFO`.

pekkaklarck commented on src/robot/reporting/jsmodelbuilders.py:
```diff
@@ -144,7 +144,7 @@ def build(self, kw, split=False):
 
     def build(self, msg):
-        if msg.level == 'WARN':
+        if msg.level == ('WARN','ERROR'):
```
Should be `in` not `==` here!! This is likely to cause unit test failures.

jaloren replied:
yikes!! Indeed fixing that stupid mistake made the unit tests pass. oui.

pekkaklarck commented:
Added few more comments. Mainly about whitespace (yes, we are pretty strict with it) but there were few possible doc enhancements too. Also noticed a bug that may be causing the unit test failures you have encountered.

pekkaklarck commented:
This looks ready to be merged. I'll proofread docs once more after that, but can do possible minor changes myself. Could you @jaloren add yourself to AUTHORS.txt and squash commits?

pekkaklarck commented:
There was a question about IDEs. My favorite Python IDE currently is PyCharm. It has several Robot plugins, and the one I've used (I think its called Intellibot) provides enough convenience that I very seldom need to use RIDE.
